File: lib/backup.py
"""AWS Backup infrastructure for disaster recovery."""

from constructs import Construct
from cdktf_cdktf_provider_aws.backup_vault import BackupVault
from cdktf_cdktf_provider_aws.backup_plan import (
    BackupPlan,
    BackupPlanRule,
    BackupPlanRuleLifecycle,
)
from cdktf_cdktf_provider_aws.backup_selection import (
    BackupSelection,
    BackupSelectionSelectionTag,
)
from cdktf_cdktf_provider_aws.iam_role import IamRole
from cdktf_cdktf_provider_aws.iam_role_policy_attachment import IamRolePolicyAttachment
import json


class BackupConstruct(Construct):
    """Construct for AWS Backup infrastructure."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_suffix: str,
        dynamodb_table_arn: str,
    ):
        """Initialize backup infrastructure."""
        super().__init__(scope, construct_id)

        # Create backup vault
        vault = BackupVault(
            self, "backup_vault", name=f"healthcare-backup-vault-{environment_suffix}"
        )

        # Create backup IAM role
        backup_role = IamRole(
            self,
            "backup_role",
            name=f"healthcare-backup-role-{environment_suffix}",
            assume_role_policy=json.dumps(
                {
                    "Version": "2012-10-17",
                    "Statement": [
                        {
                            "Effect": "Allow",
                            "Principal": {"Service": "backup.amazonaws.com"},
                            "Action": "sts:AssumeRole",
                        }
                    ],
                }
            ),
        )

        # Attach AWS Backup service role policies
        IamRolePolicyAttachment(
            self,
            "backup_policy_attachment",
            role=backup_role.name,
            policy_arn="arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForBackup",
        )

        IamRolePolicyAttachment(
            self,
            "restore_policy_attachment",
            role=backup_role.name,
            policy_arn="arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForRestores",
        )

        # Create backup plan
        plan = BackupPlan(
            self,
            "backup_plan",
            name=f"healthcare-backup-plan-{environment_suffix}",
            rule=[
                BackupPlanRule(
                    rule_name="daily_backup",
                    target_vault_name=vault.name,
                    schedule="cron(0 2 * * ? *)",  # Daily at 2 AM UTC
                    lifecycle=BackupPlanRuleLifecycle(
                        delete_after=7  # Retain for 7 days
                    ),
                )
            ],
        )

        # Create backup selection
        BackupSelection(
            self,
            "backup_selection",
            name=f"healthcare-resources-{environment_suffix}",
            iam_role_arn=backup_role.arn,
            plan_id=plan.id,
            resources=[dynamodb_table_arn],
        )

File: archive/cdktf-ts/Pr7779/lib/constructs/backup.ts
import { Construct } from 'constructs';
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';
import { BackupVault } from '@cdktf/provider-aws/lib/backup-vault';
import { BackupPlan } from '@cdktf/provider-aws/lib/backup-plan';
import { BackupSelection } from '@cdktf/provider-aws/lib/backup-selection';
import { IamRole } from '@cdktf/provider-aws/lib/iam-role';
import { IamRolePolicyAttachment } from '@cdktf/provider-aws/lib/iam-role-policy-attachment';

export interface BackupConstructProps {
  environmentSuffix: string;
  primaryProvider: AwsProvider;
  secondaryProvider: AwsProvider;
  primaryRegion: string;
  secondaryRegion: string;
}

export class BackupConstruct extends Construct {
  constructor(scope: Construct, id: string, props: BackupConstructProps) {
    super(scope, id);

    const {
      environmentSuffix,
      primaryProvider,
      secondaryProvider,
      // primaryRegion and secondaryRegion are not directly used but define vault locations via providers
    } = props;

    // Primary Region Backup Vault
    const primaryVault = new BackupVault(this, 'PrimaryBackupVault', {
      provider: primaryProvider,
      name: `dr-backup-vault-primary-${environmentSuffix}`,
      tags: {
        Name: `dr-backup-vault-primary-${environmentSuffix}`,
      },
    });

    // Secondary Region Backup Vault
    const secondaryVault = new BackupVault(this, 'SecondaryBackupVault', {
      provider: secondaryProvider,
      name: `dr-backup-vault-secondary-${environmentSuffix}`,
      tags: {
        Name: `dr-backup-vault-secondary-${environmentSuffix}`,
      },
    });

    // Backup IAM Role
    const backupRole = new IamRole(this, 'BackupRole', {
      provider: primaryProvider,
      name: `backup-role-${environmentSuffix}`,
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: {
              Service: 'backup.amazonaws.com',
            },
            Action: 'sts:AssumeRole',
          },
        ],
      }),
      tags: {
        Name: `backup-role-${environmentSuffix}`,
      },
    });

    // Attach AWS Backup service role policies
    new IamRolePolicyAttachment(this, 'BackupRolePolicy', {
      provider: primaryProvider,
      role: backupRole.name,
      policyArn:
        'arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForBackup',
    });

    new IamRolePolicyAttachment(this, 'BackupRestoreRolePolicy', {
      provider: primaryProvider,
      role: backupRole.name,
      policyArn:
        'arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForRestores',
    });

    // Backup Plan with Cross-Region Copy
    const backupPlan = new BackupPlan(this, 'BackupPlan', {
      provider: primaryProvider,
      name: `dr-backup-plan-${environmentSuffix}`,
      rule: [
        {
          ruleName: 'daily-backup',
          targetVaultName: primaryVault.name,
          schedule: 'cron(0 2 * * ? *)',
          startWindow: 60,
          completionWindow: 120,
          lifecycle: {
            deleteAfter: 7,
          },
          copyAction: [
            {
              destinationVaultArn: secondaryVault.arn,
              lifecycle: {
                deleteAfter: 7,
              },
            },
          ],
        },
      ],
      tags: {
        Name: `dr-backup-plan-${environmentSuffix}`,
      },
    });

    // Backup Selection - EBS Volumes
    new BackupSelection(this, 'BackupSelection', {
      provider: primaryProvider,
      name: `dr-backup-selection-${environmentSuffix}`,
      planId: backupPlan.id,
      iamRoleArn: backupRole.arn,
      selectionTag: [
        {
          type: 'STRINGEQUALS',
          key: 'Backup',
          value: 'true',
        },
      ],
    });
  }
}
